contracts/pools/rai/StableSwapRAI.vy

```diff
@@ -0,0 +1,1136 @@
+    @param _redemption_price_snap Address of contract providing snapshot of redemption price
+    @param _A Amplification coefficient multiplied by n * (n - 1)
+    @param _fee Fee to charge for exchanges
+    @param _admin_fee Admin fee
+    """
+    for i in range(N_COINS):
+        assert _coins[i] != ZERO_ADDRESS
+    self.coins = _coins
+    self.initial_A = _A * A_PRECISION
+    self.future_A = _A * A_PRECISION
+    self.fee = _fee
+    self.admin_fee = _admin_fee
+    self.owner = _owner
+    self.kill_deadline = block.timestamp + KILL_DEADLINE_DT
+    self.lp_token = _pool_token
+
+    self.base_pool = _base_pool
+    self.base_virtual_price = Curve(_base_pool).get_virtual_price()
+    self.base_cache_updated = block.timestamp
+    self.redemption_price_snap = _redemption_price_snap
+    for i in range(BASE_N_COINS):
+        base_coin: address = Curve(_base_pool).coins(i)
+        self.base_coins[i] = base_coin
+
+        # approve underlying coins for infinite transfers
+        response: Bytes[32] = raw_call(
+            base_coin,
+            concat(
+                method_id("approve(address,uint256)"),
+                convert(_base_pool, bytes32),
+                convert(MAX_UINT256, bytes32),
+            ),
+            max_outsize=32,
+        )
+        if len(response) > 0:
+            assert convert(response, bool)
+
+
+@view
+@internal
+def _A() -> uint256:
+    """
+    Handle ramping A up or down
+    """
+    t1: uint256 = self.future_A_time
+    A1: uint256 = self.future_A
+
+    if block.timestamp < t1:
+        A0: uint256 = self.initial_A
+        t0: uint256 = self.initial_A_time
+        # Expressions in uint256 cannot have negative numbers, thus "if"
+        if A1 > A0:
+            return A0 + (A1 - A0) * (block.timestamp - t0) / (t1 - t0)
+        else:
+            return A0 - (A0 - A1) * (block.timestamp - t0) / (t1 - t0)
+
+    else:  # when t1 == 0 or block.timestamp >= t1
+        return A1
+
+
+@view
+@external
+def A() -> uint256:
+    return self._A() / A_PRECISION
+
+
+@view
+@external
+def A_precise() -> uint256:
+    return self._A()
+
+
+@view
+@internal
+def _get_scaled_redemption_price() -> uint256:
+    """
+    @notice Reads a snapshot view of redemption price
+    @dev The fetched redemption price uses 27 decimals
+    @return The redemption price with appropriate scaling to match LP tokens vitual price.
+    """
+    rate: uint256 = RedemptionPriceSnap(self.redemption_price_snap).snappedRedemptionPrice()
+    return rate / REDMPTION_PRICE_SCALE
+
+
+@view
+@internal
+def _xp(_vp_rate: uint256) -> uint256[N_COINS]:
+    result: uint256[N_COINS] = [self._get_scaled_redemption_price(), _vp_rate]
+    for i in range(N_COINS):
+        result[i] = result[i] * self.balances[i] / PRECISION
+    return result
+
+
+@pure
+@internal
+def _xp_mem(_redemption_price: uint256, _vp_rate: uint256, _balances: uint256[N_COINS]) -> uint256[N_COINS]:
```

Comment:
Probably can pass a `rates` array instead. This is hardcoded for 2-coin pools (which is fine) but also makes it a bit more verbose. Though doesn't do anything bad also